In [14]:
#Q1
import math
import numpy as np

def single_input_neuron(p, w, b, f):
    # Calculate the output of the summation
    n = w * p + b
    
    # Apply the transfer function
    if f == 'linear':
        return n  # Linear transfer function: output is the same as the summation
    elif f == 'logsigmoid':
        return 1.0 / (1 + np.exp(-n))  # Sigmoid transfer function: output is the sigmoid of the summation
    elif f == 'relu':
        return np.where(n < 0, 0 , n)  # ReLU transfer function: output is the maximum of 0 and the summation
    elif f == 'hardlimit':
        return np.where(n >= 0, 1 , 0)  # Hardlimit transfer function: output is 1 if n is greater than or equal to 0, otherwise 0
    elif f == 'tanh':
        return (np.exp(n) - np.exp(-n)) / (np.exp(n) + np.exp(-n))  # Tanh transfer function
    else:
        raise ValueError("Invalid transfer function. Please choose 'linear', 'logsigmoid', 'relu', 'hardlimit', or 'tanh'.")

# Ask the user for inputs
p = float(input("Enter the input value (p): "))
w = float(input("Enter the weight (w): "))
b = float(input("Enter the bias (b): "))

while True:
    f = input("Enter the transfer function (linear, logsigmoid, relu, hardlimit, or tanh), or press Enter to terminate: ")
    if f == "":
        break
    
    try:
        output = single_input_neuron(p, w, b, f)
        print("Output (a):", output)
    except ValueError as e:
        print("Error:", str(e))

Enter the input value (p):  2
Enter the weight (w):  -1.5
Enter the bias (b):  3
Enter the transfer function (linear, logsigmoid, relu, hardlimit, or tanh), or press Enter to terminate:  hardlimit


Output (a): 1


Enter the transfer function (linear, logsigmoid, relu, hardlimit, or tanh), or press Enter to terminate:  relu


Output (a): 0.0


Enter the transfer function (linear, logsigmoid, relu, hardlimit, or tanh), or press Enter to terminate:  linear


Output (a): 0.0


Enter the transfer function (linear, logsigmoid, relu, hardlimit, or tanh), or press Enter to terminate:  logsigmoid


Output (a): 0.5


Enter the transfer function (linear, logsigmoid, relu, hardlimit, or tanh), or press Enter to terminate:  tanh


Output (a): 0.0


Enter the transfer function (linear, logsigmoid, relu, hardlimit, or tanh), or press Enter to terminate:  


# Check if each of the three answers that you have given above are correct? for Q1
# a=f(wp+b)
# p=2 , w=-1.5 , b=3**

----------------------------------------------------------------------------------

**1-hardlimit funcrion-->
a=hardlimit(-1.5*2+3) 
a=hardlimit(-3+3)
a=hardlimit(0) >=0 equals to 1**

---------------------------------------------------------------------------------

**2-logsigmoid function-->
a=logsigmoid(-1.5*2+3)
a=logsigmoid(-3+3)
a=logsigmoid(0) we apply the value 0 which is n to the logsigmoid function
logsigmoid=1/(1+e^-n)=1/(1+e^0)=0.5 so n=0.5 and the function =0.5**

----------------------------------------------------------------------------------


**3-poslin'Relu' function-->
poslin function-->
a=poslin(-1.5*2+3)
a=poslin(-3+3)
a=poslin(0) =0 since the condition is  n < 0, relu(n) = 0**


----------------------------------------------------------------------------------

# Q2 Develop a multiple input neuron. Assuming that it can take as input a column vector p , weight matrix w, bias vector b and produces a summation output n which when passed to the transfer function f 
# generates the output a.    

In [2]:
def neuron(p, w, b, f):
    wp = np.dot(w, p)
    n = wp + b

    if f == 'linear':
        return n  # Linear transfer function: output is the same as the summation
    elif f == 'logsigmoid':
        return 1.0 / (1 + np.exp(-n))  # Sigmoid transfer function: output is the sigmoid of the summation
    elif f == 'relu':
        return np.maximum(0, n)  # ReLU transfer function: output is the maximum of 0 and the summation
    elif f == 'hardlimit':
        return np.where(n >= 0, 1, 0)  # Hardlimit transfer function: output is 1 if n is greater than or equal to 0, otherwise 0
    elif f == 'tanh':
        return (np.exp(n) - np.exp(-n)) / (np.exp(n) + np.exp(-n))
    else:
        raise ValueError("Invalid transfer function. Please choose 'linear', 'logsigmoid', 'relu', 'hardlimit', or 'tanh'.")

p = np.array([[1], [2], [3]])  # column matrix
w = np.array([4, 5, 6])  # row matrix
b = -1.5

# Test with different transfer functions
transfer_functions = ['linear', 'logsigmoid', 'relu', 'hardlimit', 'tanh']

for f in transfer_functions:
    output = neuron(p, w, b, f)
    print("Transfer function:", f)
    print("Output (a):", float(output))
    print()

Transfer function: linear
Output (a): 30.5

Transfer function: logsigmoid
Output (a): 0.9999999999999432

Transfer function: relu
Output (a): 30.5

Transfer function: hardlimit
Output (a): 1.0

Transfer function: tanh
Output (a): 1.0



# Q2 What the output ‘a’ of the multiple input neuron if  p=[1 2 3], w=[4 5 6], b=-1.5 and transfer function is 
# ‘poslin’? 

# Check if the answer that you have given above is correct?  
# 
# positive linearl / Relu =30.5
# because
# [1]
# [2]   dot  [4,5,6]  + (-1.5)   =1*4+ 2*5 + 3*6   30.5
# [3]
# 3 x 1 and 1 x 3     where     1=1
# 
# the condition is :
# n ≥ 0, relu(n) = n
# else
# n < 0, relu(n) = 0 